designer/server/src/main/scala/pl/touk/nussknacker/ui/process/ProcessStateDefinitionService.scala

```diff
@@ -0,0 +1,79 @@
+      .groupBy { case (_, statusName, _) => statusName }
+      .map { case (statusName, stateDefinitionsByName) =>
+        // here we assume it is asserted that all definitions are unique
+        val stateDefinition = stateDefinitionsByName.map { case (_, _, sd) => sd }.head
+        val categoriesWhereStateAppears = stateDefinitionsByName
+          .flatMap { case (processingType, _, _) => categoryService.getProcessingTypeCategories(processingType) }
+        UIStateDefinition(
+          statusName,
+          stateDefinition.displayableName,
+          stateDefinition.icon,
+          stateDefinition.tooltip,
+          categoriesWhereStateAppears
+        )
+      }
+      .toList
+  }
+}
+
+// TODO: Move validation and stateDefinitions aggregation to ProcessingTypeDataReader (where processing types are reloaded).
+// Now we have only late validation when UI asks for state definitions, and state aggregation is performed here
+// every time validation is executed, but should happen only once per configuration reload.
+// That requires ProcessingTypeDataProvider.all to be an entity that combines all processing types AND
+// has all-processing-type aggregates, such as aggregated stateDefinitions.
+object ProcessStateDefinitionService {
+
+  /**
+    * Each processing type define its own state definitions. Technically it is possible that two processing types provide
+    * states with the same StatusName and different UI configurations (displayable name and icon). Here is an assertion
+    * that this does not happen and each state has the same definition across all processingTypes.
+    */
+  def validate(processingTypes: Map[ProcessingType, ProcessingTypeData]): Unit = {
+    val countDefinitionsForName = processingTypeStateDefinitions(processingTypes)
+      .groupBy { case (_, statusName, _) => statusName }
+      .map { case (statusName, stateDefinitionsForOneName) =>
+        val uniqueDefinitionsForName = stateDefinitionsForOneName
+          .groupBy { case (_, _, sd) => (sd.displayableName, sd.icon) }
+        (statusName, uniqueDefinitionsForName.size)
+      }
+    val namesWithNonUniqueDefinitions = countDefinitionsForName
+      .collect { case (statusName, size) if size > 1 => statusName }
+
+    if (namesWithNonUniqueDefinitions.nonEmpty) {
+      throw new IllegalStateException(s"State definitions are not unique for states: ${namesWithNonUniqueDefinitions.mkString(", ")}")
+    }
+  }
+
+  private def processingTypeStateDefinitions(processingTypes: Map[ProcessingType, ProcessingTypeData]): List[(ProcessingType, StatusName, StateDefinitionDetails)] = {
+    processingTypes.toList.flatMap { case (processingType, processingTypeData) =>
+      processingTypeData
+        .deploymentManager
+        .processStateDefinitionManager
+        .stateDefinitions
+        .map { case (name, sd) => (processingType, name, sd) }
+    }
+  }
+}
+
+@JsonCodec case class UIStateDefinition(name: StatusName,
```

Comment:
Do we need categories here? Why not do filtering statuses on the backend based on user permission?